helix-core/src/main/java/org/apache/helix/manager/zk/ZkBaseDataAccessor.java

```diff
@@ -92,19 +98,42 @@ public ZkBaseDataAccessor(HelixZkClient zkClient) {
     _zkClient = zkClient;
   }
 
+  private synchronized HelixZkClient getNonZNRecordClient() {
+    if (_nonZNRecordClient == null) {
```

Comment:
I think the use of synchronization primitives could be changed to allow for more parallelism?

Instead of synchronizing on the whole method, could we do:

```
private HelixZkClient getNonZNRecordZkClient() {
  if (_nonZnRecordClient == null) {
    synchronized (`_nonZnRecordZkClient`) {
      // create...
  else {
    return `_nonZnRecordZkClient`;
```

Synchronizing the whole method might significantly reduce parallelism. We only need to synchronize when we need to modify the underlying object, which is `_nonZnRecordZkClient`;

Reply:
done